timemachine/fe/bar.py

```diff
@@ -88,3 +95,76 @@ def dG_dw(w):
     dBAR_dA = jax.grad(BARzero, argnums=(1,))
     dG_dw = -dBAR_dw(w, dG)[0] / dBAR_dA(w, dG)[0]
     return dG_dw
+
+
+def bootstrap_bar(w_F, w_R, n_bootstrap=1000, timeout=10):
+    """Subsample w_F, w_R with replacement and re-run BAR many times
+
+    Parameters
+    ----------
+    w_F : array
+        forward works
+    w_R : array
+        reverse works
+    n_bootstrap : int
+        # bootstrap samples
+    timeout : int
+        in seconds
+
+    Returns
+    -------
+    best_estimate : float
+        BAR(w_F, w_R, computeUncertainty=False)
+    bootstrap_samples: array
+        length <= n_bootstrap
+        (length < n_bootstrap if timed out)
+
+    Notes
+    -----
+    * TODO[deboggle] -- upgrade from pymbar3 to pymbar4 and remove this
+    * TODO[performance] -- multiprocessing, if needed?
+    """
+    full_bar_result = pymbar.BAR(w_F, w_R, compute_uncertainty=False)
+
+    n_F, n_R = len(w_F), len(w_R)
+
+    bootstrap_samples = []
+
+    t0 = time()
+
+    for _ in range(n_bootstrap):
+        elapsed_time = time() - t0
+        if elapsed_time > timeout:
+            break
+
+        inds_F = np.random.randint(0, n_F, n_F)
+        inds_R = np.random.randint(0, n_R, n_R)
+
+        bar_result = pymbar.BAR(
+            w_F=w_F[inds_F],
+            w_R=w_R[inds_R],
+            DeltaF=full_bar_result,  # warm start
+            compute_uncertainty=False,
+            relative_tolerance=1e-6,  # reduce cost
+        )
+
+        bootstrap_samples.append(bar_result)
+
+    return full_bar_result, np.array(bootstrap_samples)
+
+
+def bar_with_bootstrapped_uncertainty(w_F, w_R, n_bootstrap=1000, timeout=10):
+    """Drop-in replacement for pymbar.BAR(w_F, w_R) -> (df, ddf)
+    where first return is forwarded from pymbar.BAR but second return is computed by bootstrapping"""
+
+    df, bootstrap_dfs = bootstrap_bar(w_F, w_R, n_bootstrap=n_bootstrap, timeout=timeout)
+
+    # warn if bootstrap distribution deviates significantly from normality
+    normaltest_result = normaltest(bootstrap_dfs)
+    pvalue_threshold = 1e-3  # arbitrary, small
+    if normaltest_result.pvalue < pvalue_threshold:
+        logger.warning(f"bootstrapped errors non-normal: {normaltest_result}")
+
+    # regardless, summarize as if normal
+    ddf = np.std(bootstrap_dfs)
```

Comment:
later on it might be useful to just return the 95% CI, but 1) I have no idea how to propagate the CI errors 2) how to use them in MLE settings, this is probably fine.

Reply:
This is why I exposed two functions: one that returns an array of bootstrap samples (which can be propagated through later steps, visualized in a histogram, or summarized directly `ci_95 = np.quantile(bootstrap_dfs, [0.025, 0.975])` if desired).

For (1) and (2), it would be most general to propagate the samples and summarize at the end, rather than to write code for propagating specific kinds of summaries such as intervals or Gaussians.

However, I think the requirements we discussed yesterday were to return Gaussian error summaries from each edge, and then use diffnet for error propagation from Gaussian errors on edges to Gaussian errors on nodes. We could also avoid ever assuming Gaussian errors, e.g. by bootstrapping the whole diffnet-included estimator as a function of the works.